support-frontend/assets/helpers/redux/commonState/reducer.ts

```diff
@@ -0,0 +1,73 @@
+import type { PayloadAction } from '@reduxjs/toolkit';
+import { createSlice } from '@reduxjs/toolkit';
+import type { ContributionTypes } from 'helpers/contributions';
+import type { ExistingPaymentMethod } from 'helpers/forms/existingPaymentMethods/existingPaymentMethods';
+import type { IsoCountry } from 'helpers/internationalisation/country';
+import { fromCountry } from 'helpers/internationalisation/countryGroup';
+import { fromCountryGroupId } from 'helpers/internationalisation/currency';
+import type { CommonStateSetupData, Internationalisation } from './state';
+import { initialCommonState } from './state';
+
+function getInternationalisationFromCountry(
+	countryId: IsoCountry,
+	internationalisation: Internationalisation,
+) {
+	const countryGroupId =
+		fromCountry(countryId) ?? internationalisation.countryGroupId;
+	const currencyId =
+		fromCountryGroupId(countryGroupId) ?? internationalisation.currencyId;
+	return {
+		countryGroupId,
+		currencyId,
+		countryId,
+	};
+}
+
+export const commonSlice = createSlice({
```

Comment:
This is the replacement for both the old switch statement reducer and action creator functions. The `state` argument passed to each reducer function is a mutable copy of the previous state object which is then replaced under the hood, enabling us to write simpler code without losing the benefits of immutable state.

Reply:
I was slightly dubious about making code that looks like it's doing mutations actually still give us an immutable state. But, it's hard to argue with how much simpler the code is.

Also removing the massive switch statement and all the manually defined action creators is amazing!

Reply:
It's a bit weird to write, but the types help make it clear what `state` here actually is. It's all based on Immer under the hood- pretty cool stuff, I'm thinking I should probably write some additions to our wiki about this https://redux-toolkit.js.org/usage/immer-reducers